Made FolderObserver a configured component instead of a ManagedService

kaikreuzer opened the pull request:
The FolderObserver had some potential issues:
- The activate method missed to call `super.activate()` (how did that actually then ever work?)
- The list of ignoredFiles was not cleared on deactivation; so when deactivating and activating the component, it was possible that it was still filled.
- The addModelParser() method directly executed code that tried to parse ignoredFiles. BUT this method was potentially called BEFORE the component was activated, which can lead to `java.lang.IllegalStateException: ScriptServiceUtil not initialized yet!` exceptions.
- The component was a ManagedService, which received its configuration only after being activated and then manually called deactivate/activate (woooaaah, looks dangerous 😲 ).

This PR includes:
- I have now changed the implementation to receive the (mandatory) configuration directly during activation, so the service does not implement `ManagedService` anymore.
- The ignoredFiles as well as other fields are cleared upon deactivation.
- `super.activate()` is correctly called.
- All models are removed from the model repo upon component deactivation.
- I removed the dynamic configuration change possibility as anyhow the list of model folders is static and there isn't a real use case where someone would want to change this list during runtime. And if it happens, DS will simply deactivate the component and re-activate it again with a new config.

I have some hope that this can actually fix #3562. @sjka should definitely have a look over it!

Signed-off-by: Kai Kreuzer <kai@openhab.org>

sjka commented on bundles/model/org.eclipse.smarthome.model.core/META-INF/MANIFEST.MF:
```diff
@@ -25,6 +25,7 @@ Import-Package: com.google.common.base,
  org.eclipse.xtext.resource.impl,
  org.osgi.framework,
  org.osgi.service.cm,
```
we don't need the cm package anymore, do we?

kaikreuzer replied:
correct, updated PR.

sjka commented:
Thanks! 
What I missed so far was indeed the fact that the problematic calls came in via `FolderObserver.addModelParser(..)`, so this really looks like it could be the yet missing piece we were searching for.

sjka commented:
However:

```
Failed tests: 
  FolderObserverTest.setUp:88->setUpServices:117 Folder Observer service cannot be found
Expected: is not null
     but: was null
  FolderObserverTest.setUp:88->setUpServices:117 Folder Observer service cannot be found
Expected: is not null
     but: was null
  FolderObserverTest.setUp:88->setUpServices:117 Folder Observer service cannot be found
Expected: is not null
     but: was null
  FolderObserverTest.setUp:88->setUpServices:117 Folder Observer service cannot be found
Expected: is not null
     but: was null
  FolderObserverTest.setUp:88->setUpServices:117 Folder Observer service cannot be found
Expected: is not null
     but: was null
  FolderObserverTest.setUp:88->setUpServices:117 Folder Observer service cannot be found
Expected: is not null
     but: was null
  FolderObserverTest.setUp:88->setUpServices:117 Folder Observer service cannot be found
Expected: is not null
     but: was null
  FolderObserverTest.setUp:88->setUpServices:117 Folder Observer service cannot be found
Expected: is not null
     but: was null
  FolderObserverTest.setUp:88->setUpServices:117 Folder Observer service cannot be found
Expected: is not null
     but: was null
  FolderObserverTest.setUp:88->setUpServices:117 Folder Observer service cannot be found
Expected: is not null
     but: was null
```

Looks like the service now is not running anmore in the tests because the required configuration is missing.
I wonder why this needs to be run as an OSGi test at all...

kaikreuzer commented:
Ok, I'll deal with the tests (didn't have them in my workspace, so I missed them...)

sjka commented:
I took the freedom to adapt the tests and pushed them here.
They got a little makeover while I ported them to Java, so that now they don't depend on the OSGi runtime anymore.

kaikreuzer commented:
> Has been this. added in front of members by intention (also if it is not necessary)?

Yeah, it was probably done rather out of a mood. I added it while refactoring, because there were two similarly named variables, one being a field, the other being a local method variable. But I agree that we do not do it in general and I could have removed it here after the code was clean again anyhow.

sjka commented:
> Should we create a parent class for JavaOSGiTest named JavaTest (some time -- not in this PR) so non-OSGi based test could use that base class?

Yes and no. It makes sense to have it - but I'm not so fond of base classes. How about a utility class and/or junit rule (depending on the purpose)?
